weave/integrations/langchain/langchain_test.py

```diff
@@ -0,0 +1,605 @@
+@pytest.mark.skip  # TODO: remove this once the langchain issue is fixed
+async def test_simple_chain_abatch(
+    client: WeaveClient,
+) -> None:
+    from langchain_core.prompts import PromptTemplate
+    from langchain_openai import ChatOpenAI
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+    prompt = PromptTemplate.from_template("1 + {number} = ")
+
+    llm_chain = prompt | llm
+    _ = await llm_chain.abatch([{"number": 2}, {"number": 3}])
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_chain_batch(res.calls)
+
+
+def assert_correct_calls_for_chain_batch_from_op(calls: list[tsi.CallSchema]) -> None:
+    assert len(calls) == 9
+    flattened = flatten_calls(calls)
+
+    got = [(op_name_from_ref(c.op_name), d, c.parent_id) for (c, d) in flattened]
+    ids = [c.id for (c, _) in flattened]
+
+    exp = [
+        ("run_batch", 0, None),
+        ("langchain.Chain.RunnableSequence", 1, ids[0]),
+        ("langchain.Prompt.PromptTemplate", 2, ids[1]),
+        ("langchain.Llm.ChatOpenAI", 2, ids[1]),
+        ("openai.chat.completions.create", 3, ids[3]),
+        ("langchain.Chain.RunnableSequence", 1, ids[0]),
+        ("langchain.Prompt.PromptTemplate", 2, ids[5]),
+        ("langchain.Llm.ChatOpenAI", 2, ids[5]),
+        ("openai.chat.completions.create", 3, ids[7]),
+    ]
+    assert got == exp
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+def test_simple_chain_batch_inside_op(
+    client: WeaveClient,
+) -> None:
+    # This test is the same as test_simple_chain_batch, but ensures things work when nested in an op
+    from langchain_core.prompts import PromptTemplate
+    from langchain_openai import ChatOpenAI
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+    prompt = PromptTemplate.from_template("1 + {number} = ")
+
+    llm_chain = prompt | llm
+
+    @weave.op()
+    def run_batch(batch: list) -> None:
+        _ = llm_chain.batch(batch)
+
+    run_batch([{"number": 2}, {"number": 3}])
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_chain_batch_from_op(res.calls)
+
+
+def assert_correct_calls_for_rag_chain(calls: list[tsi.CallSchema]) -> None:
+    assert len(calls) == 10
+    flattened = flatten_calls(calls)
+
+    got = [(op_name_from_ref(c.op_name), d) for (c, d) in flattened]
+
+    exp = [
+        ("langchain.Chain.RunnableSequence", 0),
+        ("langchain.Chain.RunnableParallel context question ", 1),
```

Comment:
I just happen to take a closer look at this RAG example, and noticed that this name looks really odd (particularly the spaces). Seems like something is wrong here

Reply:
The original `Run` is named using url unsafe characters it looks something like this `RunnableParallel<context,question>`. To make this URL safe we [replaced the non](https://github.com/wandb/weave/blob/430666487af6ad05b9eada167b74cf52d9bf6154/weave/integrations/langchain/langchain.py#L40) URL encodable characters with space. Is there a better way to do this?

Reply:
I see - ok. Can you replace it with an underscore rather than a space?

Reply:
There are some run names with `_` in them example: `chat_history` or `agent_executor`. I've replaced the space with a `-` hyphen instead.

Reply:
I see you changed this to hyphens, not underscores - why?

Reply:
the only note here is that the last path component of the op name would be nice if it was a valid python function name, which hyphens would not be. But this is not a hard rule.

Also, i don't think you need to regen all the cassettes to make this change (:

Reply:
Okay, I used `underscores` instead and updated the tests